h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -437,6 +440,132 @@ public void modelFailed(Model failedModel) {
+   * Implementation differs from {@link RandomDiscreteValueWalker} as we first eagerly evaluate the whole space, shuffle it and then iterate sequentially.
+   * Materialised grid might be taking considerable amount of memory so it needs to be taken into account.
+   */
+  public static class MaterializedRandomWalker<MP extends Model.Parameters>
+          extends BaseWalker<MP, HyperSpaceSearchCriteria.RandomDiscreteValueSearchCriteria> {
+
+    Random random;
+    ArrayList<Function<HashMap<String, Object>, HashMap<String, Object>>> _filterFunctions;
+
+    public MaterializedRandomWalker(MP params,
+                                    Map<String, Object[]> hyperParams,
+                                    ModelParametersBuilderFactory<MP> paramsBuilderFactory,
+                                    HyperSpaceSearchCriteria.RandomDiscreteValueSearchCriteria searchCriteria) {
+      this(params, hyperParams, paramsBuilderFactory, searchCriteria, null);
+    }
+    
+    public MaterializedRandomWalker(MP params,
+                                    Map<String, Object[]> hyperParams,
+                                    ModelParametersBuilderFactory<MP> paramsBuilderFactory,
+                                    HyperSpaceSearchCriteria.RandomDiscreteValueSearchCriteria searchCriteria,
+                                    ArrayList<Function<HashMap<String, Object>, HashMap<String, Object>>> filterFunctions) {
+      super(params, hyperParams, paramsBuilderFactory, searchCriteria);
+      
+      if (-1 == searchCriteria.seed())
+        random = new Random();                       
+      else
+        random = new Random(searchCriteria.seed());
+      
+      _filterFunctions = filterFunctions;
+    }
+
+    @Override
+    public HyperSpaceIterator<MP> iterator() {
+      ArrayList<HashMap<String, Object>> allElements = new ArrayList<>();
+
+      _hyperParams.forEach((hpKey, hpValues) -> {
+        if(allElements.isEmpty()) {
+          Arrays.stream(hpValues).forEach(hpValue -> {
+            HashMap<String, Object> newGridItem = new HashMap<>();
+            newGridItem.put(hpKey, hpValue);
+            allElements.add(newGridItem);
+          });
+        } else {
+          Stream<HashMap<String, Object>> expandedResult = allElements.stream().flatMap(existingItem -> {
+            return Arrays.stream(hpValues).map(hpValue -> {
+              HashMap<String, Object> clone = (HashMap<String, Object>) existingItem.clone();
+              clone.put(hpKey, hpValue);
+              return clone;
+            });
+          });
+          ArrayList<HashMap<String, Object>> collect = expandedResult.collect(Collectors.toCollection(ArrayList::new));
+          allElements.clear();
+          allElements.addAll(collect);
+        }
+      });
+      
+      Collections.shuffle(allElements, random);
+
+      if(_filterFunctions != null) {
+        // unfortunately we don't have foldLeft/foldRight methods
```

Comment:
you mean `reduce` because it looks like that's what you're trying to do here with your clean/addAll.

Reply:
@seb-h2o no, I meant fold as I'm iterating over functions but want to return grid items. Pseudocode from Scala:
```
val finterFunctions = Seq(function1, function2)
val allGridItems: Seq[HashMap[String, Object]] = Seq(gridItem1,.... gridItemN)
val filteredGridItems:Seq[HashMap[String, Object]] = filterFuntions.foldLeft(allGridItems)(case (nextFilterFunction, result) => nextFilterFunction.apply(result) ) 
```

Reply:
but `fold` is just `reduce` with a starting element, right?
I'm pretty sure this is equivalent to your fold: https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#reduce-T-java.util.function.BinaryOperator-

Reply:
No, it is not. You can do something with reduce that would be similar. But you can't implement fold reduce. https://stackoverflow.com/questions/41240414/equivalent-of-scalas-foldleft-in-java-8 . I will try to do it with reduce...this was the reason why I added comment for myself.

Reply:
Yeah, they are semantically different in their constraints, but for what you are trying to achieve, `reduce` would work, and maybe even better `collect`.

Reply:
The global issue here is that I wanted to start with not an identity element but with collection of all grid elements. These would prevent us from being able to use parallelisation 
```
_filterFunctions.parallelStream().reduce(allElements, ...)
```
as we won't be able to combine properly two filtered array lists.

But if we want to use no-parallel stream then we might be fine but only with `reduce` not `collect`. I think you can't do it with `collect` in a functional way:

First thing to point out is that  collect() is using BiConsumer and it is working only through side-effects. 
From BiConsumer's javadoc 
```
Represents an operation that accepts two input arguments and returns no
 * result.  This is the two-arity specialization of {@link Consumer}.
 * Unlike most other functional interfaces, {@code BiConsumer} is expected
 * to operate via side-effects.
```
So given the fact that BiConsumer operates only through side-effect, solution would not be any better then a current one.
```
BiConsumer<ArrayList<HashMap<String, Object>>, Function<HashMap<String, Object>, HashMap<String, Object>> >
                biConsumerApplyFunToArray = (a, b) -> {
          ArrayList<HashMap<String, Object>> filtered = a.stream().map(b::apply)
                  .filter(Objects::nonNull) // TODO we actually need only one of these actions `filter`/`distinct` depending on the type of the function 
                  .distinct()
                  .collect(Collectors.toCollection(ArrayList::new));
          allElements.clear();
          allElements.addAll(filtered);
        };

        BiConsumer<ArrayList<HashMap<String, Object>>, ArrayList<HashMap<String, Object>>> combiner = (a,b) -> {
          a.addAll(b); 
          allElements.clear();
          allElements.addAll(a);
        };

        ArrayList<HashMap<String, Object>> result = _filterFunctions.stream().collect(() -> allElements, biConsumerApplyFunToArray, combiner);
        allElements.clear();
        allElements.addAll(result);
``` 
or something similar as I did not get it working actually.


With `reduce` we can do better, but only one of the overloaded versions will work for us as here we can introduce type U ( which is similar to the fold):
```
<U> U reduce(U identity,
                 BiFunction<U, ? super T, U> accumulator,
                 BinaryOperator<U> combiner);
```
But what I don't like with this approach is that you will have to provide dummy `combiner` 
that will never be actually used ( as we can't use `.parallelStream`).
Combiner:
```  
(a, b) -> {
          return a;
        }
```
or something like this(can't put `null`).

So possible solution with `reduce` will be as follows:

```
BiFunction<ArrayList<HashMap<String, Object>>, Function<HashMap<String, Object>, HashMap<String, Object>>, ArrayList<HashMap<String, Object>> >
                biFunctionApplyFunToArray = (a, b) -> {
          return a.stream().map(b::apply)
                  .filter(Objects::nonNull)
                  .distinct()
                  .collect(Collectors.toCollection(ArrayList::new));
        };

        ArrayList<HashMap<String, Object>> result = _filterFunctions.stream().reduce(allElements, biFunctionApplyFunToArray, (a, b) -> {
          return a;
        });
        allElements.clear();
        allElements.addAll(result);
```

PS I would not say that `reduce` with identity element is a reduce. It is closer to fold but not exactly fold. We are fine here as we don't need combiner for parallelisation. If we needed parallelisation we are on one hand lucky that we use monoid like type ArrayList( integers for example would not combine) but on the other hand unlucky that we don't want to start with empty identity element.

Reply:
@deil87 just noticed this:

> _filterFunctions.parallelStream().reduce(allElements, ...)

I don't really think we can use "parallel" in Java because it uses a different Fork/Join pool than H2O. Just thought I would warn you early :)

BTW - why do we need to materialize all the parameters? Is that somehow required by the filtering function?

Reply:
@michalkurka 
> I don't really think we can use "parallel" in Java because it uses a different Fork/Join pool than H2O

Thanks, I was just talking in general about `fold `and `reduce`. Was not going to use `.parallelStream` in H2O.

> BTW - why do we need to materialize all the parameters? Is that somehow required by the filtering function?

I understand your interest here. There are couple of points from my side:

1) I don't think it is a requirement. But I was thinking about performance benefits. I would say that under certain conditions sequential evaluation on the whole collection of grid items is more efficient(than one by one) as inputs to filtering functions will be reducing in sizes towards the end of the collection of functions. 
2) We need grid to be materialised to be able to apply Bayesian approaches( e.g. Sequential model based optimisation) in the near future (probably mentioned t his point before). And overall I think that random selection of indices is inefficient way of traversing a space. In theory it might take forever to visit all possible permutations.  With materialised version we don't need to 1)compute hashes of indices, 2) check whether current combination has already been visited 3) store hashes in memory( not a big deal of memory required though. btw we can use HashSet for visited permutations instead of LinkedHashSet as we don't use what `Linked` is providing). Micro-benchmarking proves that materialised version is faster.  But yes, it will take more space( actually not more but just we will be keeping more at the same time). 
Ideally, as it always the case with trade-off situations, we should decide dynamically - if the grid is big then lazy evaluation if it is relatively small then we better work with materialised version.

Summary:
1) There is no hard dependency between materialisation and filtering
2) We will need Materialised version of the grid for SMBO
3) Filtering could and should be added to all Walkers. This PR was just an example on how it could be implemented for a particular walker. 

I think that I will split this PR into two:
1) Filtering support for existing BaseWalkers 
2) MaterialisedWalker(with filtering support as well)

Reply:
Forgot to mention one point that made me favour materialised version - it was easier to implement "shadowing" of hyperparameters. When you evaluate one combination from the subgroup and you need to mark all the other from this subgroup as redundant. But it could be implemented without materialisation as well.